# Data Preparation

In [1]:
# Import necessary libraries

%matplotlib inline

import pandas as pd
import numpy as np
import geopandas as gpd
import matplotlib.pyplot as plt
import plotly.express as px

import plotly.io as pio
pio.renderers.default = 'notebook'

# Suppress warnings
import warnings
warnings.filterwarnings("ignore")

In [70]:
# Load data

kidnap_count = pd.read_csv('/home/muhammad/global-kidnap/Data/National Kidnapping.csv')
kidnap_rate = pd.read_csv('./Data/National Kidnap Rate.csv')
population = pd.read_csv('./Data/Population.csv')
gdp = pd.read_csv('./Data/GDP.csv')

# Check the first five rows of the life expectancy dataset
kidnap_count.head()

location                                variable    Unit  Date  Value
0   Rwanda  Kidnapping at the National Level Count  Number  2008      8
1   Rwanda  Kidnapping at the National Level Count  Number  2009      8
2   Rwanda  Kidnapping at the National Level Count  Number  2010     12
3   Rwanda  Kidnapping at the National Level Count  Number  2011     12
4   Rwanda  Kidnapping at the National Level Count  Number  2012     10

In [71]:
# Check the number of rows

kidnap_count.shape

(1435, 5)

From the above, we found that there are **266** rows (records) and **66** attributes (columns) in df_lif_exp dataframe

In [72]:
# Check the first 5 rows in the df_popu dataframe

kidnap_rate.head()

location                               variable  \
0   Rwanda  Kidnapping at the National Level Rate   
1   Rwanda  Kidnapping at the National Level Rate   
2   Rwanda  Kidnapping at the National Level Rate   
3   Rwanda  Kidnapping at the National Level Rate   
4   Rwanda  Kidnapping at the National Level Rate   

                           Unit  Date  Value  
0  rates per 100,000 population  2008  0.084  
1  rates per 100,000 population  2009  0.082  
2  rates per 100,000 population  2010  0.120  
3  rates per 100,000 population  2011  0.117  
4  rates per 100,000 population  2012  0.095

In [73]:
# Check the number of rows

kidnap_rate.shape

(1435, 5)

In [74]:
population.head()

country             series    Unit  Date  Value
0  American Samoa  Population, total  Number  1960  20127
1  American Samoa  Population, total  Number  1961  20605
2  American Samoa  Population, total  Number  1962  21246
3  American Samoa  Population, total  Number  1963  22029
4  American Samoa  Population, total  Number  1964  22850

In [75]:
population.shape

(13717, 5)

In [76]:
gdp.head()

country                   series               Unit  Date  \
0  Congo, Dem. Rep.  GDP (constant 2010 US$)  constant 2010 US$  1970   
1  Congo, Dem. Rep.  GDP (constant 2010 US$)  constant 2010 US$  1990   
2  Congo, Dem. Rep.  GDP (constant 2010 US$)  constant 2010 US$  2000   
3  Congo, Dem. Rep.  GDP (constant 2010 US$)  constant 2010 US$  2005   
4  Congo, Dem. Rep.  GDP (constant 2010 US$)  constant 2010 US$  2010   

          Value  
0  2.582914e+10  
1  2.940473e+10  
2  1.652806e+10  
3  1.992385e+10  
4  2.609358e+10

In [77]:
gdp.shape

(2733, 5)

From the above, we found that there are **266** rows (records) and **66** attributes (columns) in df_popu dataframe

In [78]:
# Check for available datasets in the GeoPandas dataset function

gpd.datasets.available

['naturalearth_lowres', 'naturalearth_cities', 'nybb']

In [79]:
# Read in the 'naturalearth_lowres' dataset

world = gpd.read_file(gpd.datasets.get_path('naturalearth_lowres'))
world.head()

pop_est      continent                      name iso_a3  gdp_md_est  \
0     920938        Oceania                      Fiji    FJI      8374.0   
1   53950935         Africa                  Tanzania    TZA    150600.0   
2     603253         Africa                 W. Sahara    ESH       906.5   
3   35623680  North America                    Canada    CAN   1674000.0   
4  326625791  North America  United States of America    USA  18560000.0   

                                            geometry  
0  MULTIPOLYGON (((180.00000 -16.06713, 180.00000...  
1  POLYGON ((33.90371 -0.95000, 34.07262 -1.05982...  
2  POLYGON ((-8.66559 27.65643, -8.66512 27.58948...  
3  MULTIPOLYGON (((-122.84000 49.00000, -122.9742...  
4  MULTIPOLYGON (((-122.84000 49.00000, -120.0000...

In [80]:
# Check the number of rows

world.shape

(177, 6)

From the above, we found that there are **177** rows (records) and **6** attributes (columns) in the 'world' dataframe

In [81]:
# Create a new dataframe from 'world' picking only ['name', 'iso_a3', 'geometry']

df_world = world[['name', 'iso_a3', 'geometry', 'continent']]
df_world.head()

name iso_a3  \
0                      Fiji    FJI   
1                  Tanzania    TZA   
2                 W. Sahara    ESH   
3                    Canada    CAN   
4  United States of America    USA   

                                            geometry      continent  
0  MULTIPOLYGON (((180.00000 -16.06713, 180.00000...        Oceania  
1  POLYGON ((33.90371 -0.95000, 34.07262 -1.05982...         Africa  
2  POLYGON ((-8.66559 27.65643, -8.66512 27.58948...         Africa  
3  MULTIPOLYGON (((-122.84000 49.00000, -122.9742...  North America  
4  MULTIPOLYGON (((-122.84000 49.00000, -120.0000...  North America

In [82]:
# Rename 'iso_a3' and 'geometry' columns

df_world.rename(columns={'name':'country', 'iso_a3':'Country Code', 'geometry':'Geometry', 'continent':'Continent'}, inplace=True)

In [83]:
df_world = df_world[['country', 'Continent']]
df_world.head()

country      Continent
0                      Fiji        Oceania
1                  Tanzania         Africa
2                 W. Sahara         Africa
3                    Canada  North America
4  United States of America  North America

In [90]:
# Merge the world dataset on life expectany. This picks only countries (in the life expectancy dataframe) 
# that have matching country code (in the world dataframe)

kidnap = kidnap_rate.merge(kidnap_count, on=['location', 'Date'])
kidnap.head()

location                             variable_x  \
0   Rwanda  Kidnapping at the National Level Rate   
1   Rwanda  Kidnapping at the National Level Rate   
2   Rwanda  Kidnapping at the National Level Rate   
3   Rwanda  Kidnapping at the National Level Rate   
4   Rwanda  Kidnapping at the National Level Rate   

                         Unit_x  Date  Value_x  \
0  rates per 100,000 population  2008    0.084   
1  rates per 100,000 population  2009    0.082   
2  rates per 100,000 population  2010    0.120   
3  rates per 100,000 population  2011    0.117   
4  rates per 100,000 population  2012    0.095   

                               variable_y  Unit_y  Value_y  
0  Kidnapping at the National Level Count  Number        8  
1  Kidnapping at the National Level Count  Number        8  
2  Kidnapping at the National Level Count  Number       12  
3  Kidnapping at the National Level Count  Number       12  
4  Kidnapping at the National Level Count  Number       10

In [91]:
kidnap = kidnap[['location', 'Date', 'Value_x', 'Value_y']]
kidnap.rename(columns={'location':'country', 'Value_x':'Kidnapping Rate', 'Value_y':'Kidnapping Count'}, inplace=True)        
kidnap.head()

country  Date  Kidnapping Rate  Kidnapping Count
0  Rwanda  2008            0.084                 8
1  Rwanda  2009            0.082                 8
2  Rwanda  2010            0.120                12
3  Rwanda  2011            0.117                12
4  Rwanda  2012            0.095                10

In [92]:
kidnap = kidnap.merge(population, on=['country', 'Date'])
kidnap.head()

country  Date  Kidnapping Rate  Kidnapping Count             series    Unit  \
0  Rwanda  2008            0.084                 8  Population, total  Number   
1  Rwanda  2009            0.082                 8  Population, total  Number   
2  Rwanda  2010            0.120                12  Population, total  Number   
3  Rwanda  2011            0.117                12  Population, total  Number   
4  Rwanda  2012            0.095                10  Population, total  Number   

      Value  
0   9524532  
1   9782770  
2  10039338  
3  10293333  
4  10549668

In [93]:
kidnap = kidnap.merge(gdp, on=['country', 'Date'])
kidnap.head()

country  Date  Kidnapping Rate  Kidnapping Count           series_x  \
0   Rwanda  2010            0.120                12  Population, total   
1   Rwanda  2011            0.117                12  Population, total   
2   Rwanda  2012            0.095                10  Population, total   
3   Rwanda  2013            0.194                21  Population, total   
4  Algeria  2005            0.305               101  Population, total   

   Unit_x   Value_x                 series_y             Unit_y       Value_y  
0  Number  10039338  GDP (constant 2010 US$)  constant 2010 US$  6.015646e+09  
1  Number  10293333  GDP (constant 2010 US$)  constant 2010 US$  6.494395e+09  
2  Number  10549668  GDP (constant 2010 US$)  constant 2010 US$  7.055609e+09  
3  Number  10811538  GDP (constant 2010 US$)  constant 2010 US$  7.388623e+09  
4  Number  33149720  GDP (constant 2010 US$)  constant 2010 US$  1.243809e+11

In [94]:
kidnap = kidnap[['country', 'Date', 'Kidnapping Rate', 'Kidnapping Count', 'Value_x', 'Value_y']]
kidnap.rename(columns={'Value':'Population', 'Date':'Year', 'Value_x':'Population', 'Value_y':'GDP'}, inplace=True)
kidnap.head()

country  Year  Kidnapping Rate  Kidnapping Count  Population           GDP
0   Rwanda  2010            0.120                12    10039338  6.015646e+09
1   Rwanda  2011            0.117                12    10293333  6.494395e+09
2   Rwanda  2012            0.095                10    10549668  7.055609e+09
3   Rwanda  2013            0.194                21    10811538  7.388623e+09
4  Algeria  2005            0.305               101    33149720  1.243809e+11

In [95]:
kidnap.shape

(780, 6)

In [96]:
# Merge the world dataset on life expectany. This picks only countries (in the population dataframe) 
# that have matching country code (in the world dataframe)

kidnap = df_world.merge(kidnap, on='country')
kidnap.head()

country      Continent  Year  Kidnapping Rate  Kidnapping Count  Population  \
0  Canada  North America  2005           12.113              3896    32243753   
1  Canada  North America  2010           12.595              4301    34004889   
2  Canada  North America  2011           10.944              3780    34339328   
3  Canada  North America  2012           10.415              3637    34714222   
4  Canada  North America  2013            9.154              3231    35082954   

            GDP  
0  1.243745e+12  
1  1.399482e+12  
2  1.443522e+12  
3  1.468960e+12  
4  1.503174e+12

In [97]:
#kidnap['kidnap_%_Change'] = kidnap['Kidnapping Count'].pct_change()
kidnap['kidnap_rate_%_Change'] = kidnap['Kidnapping Rate'].pct_change()
kidnap.head()

country      Continent  Year  Kidnapping Rate  Kidnapping Count  Population  \
0  Canada  North America  2005           12.113              3896    32243753   
1  Canada  North America  2010           12.595              4301    34004889   
2  Canada  North America  2011           10.944              3780    34339328   
3  Canada  North America  2012           10.415              3637    34714222   
4  Canada  North America  2013            9.154              3231    35082954   

            GDP  kidnap_rate_%_Change  
0  1.243745e+12                   NaN  
1  1.399482e+12              0.039792  
2  1.443522e+12             -0.131084  
3  1.468960e+12             -0.048337  
4  1.503174e+12             -0.121075

For data consistency, we will drop the year **2005** from our dataframe

In [105]:
kidnap = kidnap[kidnap['Year'].between(2010, 2018)]
kidnap.head()

country      Continent  Year  Kidnapping Rate  Kidnapping Count  Population  \
1  Canada  North America  2010           12.595              4301    34004889   
2  Canada  North America  2011           10.944              3780    34339328   
3  Canada  North America  2012           10.415              3637    34714222   
4  Canada  North America  2013            9.154              3231    35082954   
5  Canada  North America  2014            9.225              3290    35437435   

            GDP  kidnap_rate_%_Change  
1  1.399482e+12              0.039792  
2  1.443522e+12             -0.131084  
3  1.468960e+12             -0.048337  
4  1.503174e+12             -0.121075  
5  1.546316e+12              0.007756

In [106]:
# Check for the number of rows

kidnap.shape

(604, 8)

In [107]:
kidnap.isnull().sum()

country                  0
Continent                0
Year                     0
Kidnapping Rate          0
Kidnapping Count         0
Population               0
GDP                      0
kidnap_rate_%_Change    16
dtype: int64

In [108]:
kidnap.nunique(dropna=True)

country                  89
Continent                 6
Year                      9
Kidnapping Rate         482
Kidnapping Count        305
Population              604
GDP                     604
kidnap_rate_%_Change    543
dtype: int64

In [132]:
kidnap.nlargest(5, 'Kidnapping Rate')

country Continent  Year  Kidnapping Rate  Kidnapping Count  Population  \
225  Lebanon      Asia  2016         177.7355             10460     6714281   
432   Turkey      Asia  2014          42.6690             32953    77229262   
431   Turkey      Asia  2013          41.0880             31196    75925454   
222  Lebanon      Asia  2013          20.5310              1214     5913016   
428   Turkey      Asia  2010          18.7840             13586    72326992   

              GDP  kidnap_rate_%_Change  
225  5.070321e+10             10.553270  
432  8.147437e+11              0.038478  
431  7.763921e+11              1.283682  
222  4.863679e+10              0.249376  
428  6.141711e+11              0.459859

In [133]:
kidnap.nsmallest(5, 'Kidnapping Rate')

country      Continent  Year  Kidnapping Rate  Kidnapping Count  \
155  Guyana  South America  2012              0.0                 0   
156  Guyana  South America  2013              0.0                 0   
157  Guyana  South America  2015              0.0                 0   
158  Guyana  South America  2016              0.0                 0   
159  Guyana  South America  2017              0.0                 0   

     Population           GDP  kidnap_rate_%_Change  
155      755388  4.032798e+09                  -1.0  
156      759281  4.180128e+09                   NaN  
157      767433  4.279840e+09                   NaN  
158      771363  4.442795e+09                   NaN  
159      775218  4.608711e+09                   NaN

In [109]:
kidnap.corr()

Year  Kidnapping Rate  Kidnapping Count  Population  \
Year                  1.000000         0.009828         -0.044942   -0.071548   
Kidnapping Rate       0.009828         1.000000          0.277231    0.033325   
Kidnapping Count     -0.044942         0.277231          1.000000    0.768020   
Population           -0.071548         0.033325          0.768020    1.000000   
GDP                  -0.012369         0.051020          0.208200    0.346715   
kidnap_rate_%_Change  0.045924         0.026771          0.020061   -0.005499   

                           GDP  kidnap_rate_%_Change  
Year                 -0.012369              0.045924  
Kidnapping Rate       0.051020              0.026771  
Kidnapping Count      0.208200              0.020061  
Population            0.346715             -0.005499  
GDP                   1.000000             -0.034269  
kidnap_rate_%_Change -0.034269              1.000000

In [111]:
fig = px.bar(kidnap, x='Kidnapping Rate', y='country',
             hover_data=['Population', 'GDP',],  
             color='Kidnapping Count', animation_frame='Year', template='plotly_dark', color_continuous_scale='Viridis',
             labels={'Percentage Change':'Percentage Change in Life Expectancy'}, height=400)
fig.show()

In [112]:
fig = px.bar(kidnap, x='Kidnapping Rate', y='Continent',
             hover_data=['Population', ],  
             color='Kidnapping Count', animation_frame='Year', template='plotly_dark', color_continuous_scale='Viridis',
             labels={'Percentage Change':'Percentage Change in Life Expectancy'}, height=400)
fig.show()

In [114]:
# Plot a scatter plot of the average life expectancy for each country between 2000 and 2010 
# (taking the average for each country over the years)

fig = px.scatter(kidnap, x='Kidnapping Count', y='Population', color='country', animation_frame='Year',
                 hover_data=['Kidnapping Rate', 'Continent', 'GDP', ])
fig.update_layout(title='Scatter Plot of Population VS Number of Kidnaps from year 2003 to 2018',
                  template="plotly_dark")
fig.show()

In [134]:
# Choropleth Map of Life Expectancy for Countries with Year ranging from 2000 to 2010
fig = px.choropleth(kidnap,locations='country',locationmode='country names', 
                    color='Kidnapping Rate', color_continuous_scale='Viridis', animation_frame='Year',
                    hover_data=['Continent', 'Kidnapping Count', 'Population', 'GDP',])
fig.update_layout(title='Choropleth Map of Life Expectancy for Countries with Year ranging from 2000 to 2010',
                  template="plotly_dark")
fig.show()

In [123]:
# Create a new dataframe containing the average Population, Kidnapping Count, Kidnapping Rate, etc. 
# over the years under study

df_kidnap = kidnap[['country', 'Continent', 'Kidnapping Rate', 'Kidnapping Count', 'Population', 'GDP',
                    'kidnap_rate_%_Change',]].groupby('country',
                as_index=False).mean()
df_kidnap.head()

country  Kidnapping Rate  Kidnapping Count    Population           GDP  \
0    Albania         0.149000          4.333333  2.888828e+06  1.134435e+10   
1    Algeria         0.663833        254.500000  3.780244e+07  1.527123e+11   
2    Armenia         1.536556         44.777778  2.911813e+06  1.017038e+10   
3  Australia         2.399556        562.333333  2.347783e+07  1.320698e+12   
4    Austria         0.055778          4.777778  8.580963e+06  3.825631e+11   

   kidnap_rate_%_Change  
0              0.421225  
1              0.367784  
2              0.159965  
3             -0.125083  
4              0.312303

In [124]:
df_kidnap.nlargest(5, 'Kidnapping Rate')

country  Kidnapping Rate  Kidnapping Count    Population           GDP  \
46   Lebanon        38.684071       2265.285714  5.873390e+06  4.813610e+10   
83    Turkey        19.945109      23346.571429  7.599047e+07  7.658976e+11   
7    Belgium        10.458500       1169.625000  1.117375e+07  4.527679e+11   
16    Canada        10.370875       3673.125000  3.524206e+07  1.513743e+12   
61  Pakistan         9.845889      19182.000000  1.955337e+08  2.631779e+11   

    kidnap_rate_%_Change  
46              1.878094  
83              0.083673  
7               0.022362  
16             -0.017528  
61             15.736113

In [125]:
df_kidnap.nsmallest(5, 'Kidnapping Rate')

country  Kidnapping Rate  Kidnapping Count    Population  \
72      Senegal         0.014000          2.000000  1.457845e+07   
65  Philippines         0.021333         20.333333  9.558316e+07   
49   Madagascar         0.023167          5.500000  2.267125e+07   
54      Myanmar         0.024143         12.428571  5.228169e+07   
25      Finland         0.032444          1.777778  5.451628e+06   

             GDP  kidnap_rate_%_Change  
72  1.777477e+10             -0.854167  
65  2.399613e+11              0.158942  
49  1.055317e+10                   NaN  
54  5.685279e+10                   NaN  
25  2.391304e+11              0.572403

In [127]:
fig = px.bar(df_kidnap, x='Kidnapping Rate', y='country',
             hover_data=['Kidnapping Count', 'Population', 'GDP', 'kidnap_rate_%_Change',] ,  
             color='Kidnapping Count', template='plotly_dark', color_continuous_scale='Viridis',
             labels={'Percentage Change':'Percentage Change in Life Expectancy'}, height=400)
fig.show()

In [128]:
# Plot a scatter plot of the average life expectancy for each country between 2000 and 2010 
# (taking the average for each country over the years)

fig = px.scatter(df_kidnap, x='Kidnapping Count', y='Population', color='country', color_continuous_scale='Viridis',
                 hover_data=['Kidnapping Rate', 'GDP', 'kidnap_rate_%_Change'])
fig.update_layout(title='Scatter Plot of Average Population VS Average Number of Kidnaps from year 2003 to 2018',
                  template="plotly_dark")
fig.show()

In [129]:
df_kidnap.corr()

Kidnapping Rate  Kidnapping Count  Population       GDP  \
Kidnapping Rate              1.000000          0.310092    0.054757  0.090444   
Kidnapping Count             0.310092          1.000000    0.862714  0.256232   
Population                   0.054757          0.862714    1.000000  0.350309   
GDP                          0.090444          0.256232    0.350309  1.000000   
kidnap_rate_%_Change         0.054588         -0.010829   -0.030593 -0.068336   

                      kidnap_rate_%_Change  
Kidnapping Rate                   0.054588  
Kidnapping Count                 -0.010829  
Population                       -0.030593  
GDP                              -0.068336  
kidnap_rate_%_Change              1.000000

## Africa

In [135]:
africa = kidnap[kidnap['Continent'] == 'Africa']
africa.head()

country Continent  Year  Kidnapping Rate  Kidnapping Count  Population  \
41   Kenya    Africa  2010            0.188                79    42030684   
42   Kenya    Africa  2011            0.127                55    43178270   
43   Kenya    Africa  2012            0.169                75    44343469   
44   Kenya    Africa  2013            0.130                59    45519986   
45   Kenya    Africa  2014            0.071                33    46700063   

             GDP  kidnap_rate_%_Change  
41  4.891939e+10             -0.546988  
42  5.190751e+10             -0.324468  
43  5.427616e+10              0.330709  
44  5.746688e+10             -0.230769  
45  6.054545e+10             -0.453846

In [136]:
africa.nunique(dropna=True)

country                 14
Continent                1
Year                     9
Kidnapping Rate         65
Kidnapping Count        57
Population              65
GDP                     65
kidnap_rate_%_Change    65
dtype: int64

In [139]:
africa.nlargest(5, 'Kidnapping Rate')

country Continent  Year  Kidnapping Rate  Kidnapping Count  \
56   South Africa    Africa  2017            9.569              5455   
54   South Africa    Africa  2015            9.051              5013   
55   South Africa    Africa  2016            8.880              4991   
53   South Africa    Africa  2014            8.602              4692   
194         Benin    Africa  2017            4.743               530   

     Population           GDP  kidnap_rate_%_Change  
56     57009751  3.233988e+11              0.077590  
54     55386369  3.176205e+11              0.052197  
55     56207649  3.188881e+11             -0.018893  
53     54544184  3.138737e+11             25.714286  
194    11175192  1.243594e+10            337.785714

In [140]:
africa.nsmallest(5, 'Kidnapping Rate')

country Continent  Year  Kidnapping Rate  Kidnapping Count  \
226  Madagascar    Africa  2010            0.000                 0   
228  Madagascar    Africa  2012            0.004                 1   
227  Madagascar    Africa  2011            0.005                 1   
193     Senegal    Africa  2015            0.014                 2   
48        Kenya    Africa  2018            0.021                11   

     Population           GDP  kidnap_rate_%_Change  
226    21151640  9.925208e+09             -1.000000  
228    22346641  1.038545e+10             -0.200000  
227    21743970  1.008187e+10                   inf  
193    14578450  1.777477e+10             -0.854167  
48     51392570  7.551543e+10             -0.723684

In [148]:
africa.corr()

Year  Kidnapping Rate  Kidnapping Count  Population  \
Year                  1.000000         0.412193          0.330102    0.001841   
Kidnapping Rate       0.412193         1.000000          0.927787    0.077148   
Kidnapping Count      0.330102         0.927787          1.000000    0.253968   
Population            0.001841         0.077148          0.253968    1.000000   
GDP                   0.119535         0.450875          0.628340    0.851651   
kidnap_rate_%_Change  0.207806         0.211661          0.027477   -0.085465   

                           GDP  kidnap_rate_%_Change  
Year                  0.119535              0.207806  
Kidnapping Rate       0.450875              0.211661  
Kidnapping Count      0.628340              0.027477  
Population            0.851651             -0.085465  
GDP                   1.000000             -0.062598  
kidnap_rate_%_Change -0.062598              1.000000

In [141]:
fig = px.bar(africa, x='Kidnapping Rate', y='country',
             hover_data=['Kidnapping Count', 'Population', 'GDP',],  
             color='Kidnapping Count', template='plotly_dark', color_continuous_scale='Viridis',
             animation_frame='Year',
             labels={'Percentage Change':'Percentage Change in Life Expectancy'}, height=400)
fig.show()

In [142]:
# Choropleth Map of Life Expectancy for Countries with Year ranging from 2000 to 2010
fig = px.choropleth(africa,locations='country',locationmode='country names', 
                    color='Kidnapping Count', color_continuous_scale='Viridis', animation_frame='Year',
                    hover_data=['Kidnapping Rate', 'Population', 'GDP',])
fig.update_layout(title='Choropleth Map of Life Expectancy for Countries with Year ranging from 2000 to 2010',
                  template="plotly_dark")
fig.show()

In [147]:
# Plot a scatter plot of the average life expectancy for each country between 2000 and 2010 
# (taking the average for each country over the years)

fig = px.scatter(africa, x='Kidnapping Count', y='Population', color='country', color_continuous_scale='Viridis', 
                 animation_frame='Year', hover_data=['Kidnapping Rate', 'GDP', 'kidnap_rate_%_Change'])
fig.update_layout(title='Scatter Plot of Average Population VS Average Number of Kidnaps from year 2010 to 2018',
                  template="plotly_dark")
fig.show()

In [153]:
# Create a new dataframe containing the average life expectancy for each country between 2000 and 2010

df_africa = africa[['country', 'Continent', 'Kidnapping Rate', 'Kidnapping Count', 'Population',
                    'GDP', 'kidnap_rate_%_Change']].groupby('country',
                as_index=False).mean()
df_africa.rename(columns={'Kidnapping Rate':'Average Kidnapping Rate', 'Kidnapping Count':'Average Kidnapping Count', 
                          'Population':'Average Population', 'GDP':'Average GDP', 
                          'kidnap_rate_%_Change':'Average kidnap_rate_%_Change'}, inplace=True)
df_africa.head()

country  Average Kidnapping Rate  Average Kidnapping Count  \
0   Algeria                 0.663833                254.500000   
1     Benin                 4.743000                530.000000   
2  Botswana                 0.165000                  3.333333   
3   Burundi                 0.552800                 51.800000   
4  Cameroon                 2.817250                636.875000   

   Average Population   Average GDP  Average kidnap_rate_%_Change  
0        3.780244e+07  1.527123e+11                      0.367784  
1        1.117519e+07  1.243594e+10                    337.785714  
2        2.030377e+06  1.273339e+10                     -0.182121  
3        9.252921e+06  2.965764e+09                      0.157974  
4        2.241038e+07  2.864334e+10                      0.910847

In [154]:
df_africa.nlargest(5, 'Average Kidnapping Rate')

country  Average Kidnapping Rate  Average Kidnapping Count  \
12   South Africa                 9.025500                  5037.750   
1           Benin                 4.743000                   530.000   
4        Cameroon                 2.817250                   636.875   
8         Morocco                 2.065333                   717.000   
5   Guinea-Bissau                 1.548000                    26.250   

    Average Population   Average GDP  Average kidnap_rate_%_Change  
12        5.578699e+07  3.184453e+11                      6.456295  
1         1.117519e+07  1.243594e+10                    337.785714  
4         2.241038e+07  2.864334e+10                      0.910847  
8         3.472631e+07  1.007394e+11                      0.234673  
5         1.715083e+06  1.031991e+09                      2.167189

In [155]:
df_africa.nsmallest(5, 'Average Kidnapping Rate')

country  Average Kidnapping Rate  Average Kidnapping Count  \
11     Senegal                 0.014000                  2.000000   
7   Madagascar                 0.023167                  5.500000   
6        Kenya                 0.126500                 57.500000   
10      Rwanda                 0.131500                 13.750000   
2     Botswana                 0.165000                  3.333333   

    Average Population   Average GDP  Average kidnap_rate_%_Change  
11        1.457845e+07  1.777477e+10                     -0.854167  
7         2.267125e+07  1.055317e+10                           NaN  
6         4.640807e+07  6.045827e+10                     -0.047397  
10        1.042347e+07  6.738568e+09                     -0.030053  
2         2.030377e+06  1.273339e+10                     -0.182121

In [156]:
df_africa.corr()

Average Kidnapping Rate  \
Average Kidnapping Rate                      1.000000   
Average Kidnapping Count                     0.897064   
Average Population                           0.037190   
Average GDP                                  0.378259   
Average kidnap_rate_%_Change                 0.364456   

                              Average Kidnapping Count  Average Population  \
Average Kidnapping Rate                       0.897064            0.037190   
Average Kidnapping Count                      1.000000            0.251659   
Average Population                            0.251659            1.000000   
Average GDP                                   0.604992            0.889340   
Average kidnap_rate_%_Change                 -0.005952           -0.161968   

                              Average GDP  Average kidnap_rate_%_Change  
Average Kidnapping Rate          0.378259                      0.364456  
Average Kidnapping Count         0.604992                     -0.005952  
Average Population               0.889340                     -0.161968  
Average GDP                      1.000000                     -0.166758  
Average kidnap_rate_%_Change    -0.166758                      1.000000

In [157]:
fig = px.bar(df_africa, x='Average Kidnapping Rate', y='country',
             hover_data=['Average Kidnapping Count', 'Average Population', 'Average GDP', ],  
             color='Average Kidnapping Count', template='plotly_dark', color_continuous_scale='Viridis',
             labels={'Percentage Change':'Percentage Change in Life Expectancy'}, height=400)
fig.show()

In [158]:
# Choropleth Map of Life Expectancy for Countries with Year ranging from 2000 to 2010
fig = px.choropleth(df_africa,locations='country',locationmode='country names', 
                    color='Average Kidnapping Count', color_continuous_scale='Viridis', 
                    hover_data=['Average Kidnapping Rate', 'Average Population', 'Average GDP'])
fig.update_layout(title='Choropleth Map of Life Expectancy for Countries with Year ranging from 2000 to 2010',
                  template="plotly_dark")
fig.show()

## Asia

In [54]:
asia = kidnap[kidnap['Continent']=='Asia']
asia.head()

country Continent  Year  Kidnapping Rate  Kidnapping Count  Population  \
15  Kazakhstan      Asia  2005            0.597                92    15147029   
16  Kazakhstan      Asia  2006            0.540                84    15308085   
17  Kazakhstan      Asia  2007            0.535                84    15484192   
18  Kazakhstan      Asia  2008            0.561                89    15776938   
19  Kazakhstan      Asia  2009            0.748               120    16092822   

    kidnap_%_Change  kidnap_rate_%_Change  
15        -0.975648             -0.941954  
16        -0.086957             -0.095477  
17         0.000000             -0.009259  
18         0.059524              0.048598  
19         0.348315              0.333333

In [55]:
asia.nunique(dropna=True)

country                  30
Continent                 1
Year                     16
Kidnapping Rate         223
Kidnapping Count        166
Population              257
kidnap_%_Change         217
kidnap_rate_%_Change    232
dtype: int64

## Europe

In [64]:
# Create a new dataframe for countires having population greater than 295,516,599

europe = kidnap[kidnap['Continent']=='Europe']
europe.head()

country Continent  Year  Kidnapping Rate  Kidnapping Count  Population
235  France    Europe  2003            3.371              2031    62244880
236  France    Europe  2004            3.531              2143    62704901
237  France    Europe  2005            3.292              2012    63179356
238  France    Europe  2006            3.720              2288    63621376
239  France    Europe  2007            3.409              2109    64016227

In [65]:
# Choropleth Map of Life Expectancy for Countries with Year ranging from 2000 to 2010
fig = px.choropleth(europe,locations='country',locationmode='country names', 
                    color='Kidnapping Count', color_continuous_scale='Viridis', animation_frame='Year',
                    hover_data=['Kidnapping Rate', 'Population', 'Continent'])
fig.update_layout(title='Choropleth Map of Life Expectancy for Countries with Year ranging from 2000 to 2010',
                  template="simple_white")
fig.show()

## Task 1.2.3: 

In [184]:
population_2005_3 = df_merged_data[df_merged_data['Population'].between(63179356, 1147609924)]
population_2005_3.head()

Country Name   Population  Life Expectancy
3        United States  295516599.0           75.000
7            Indonesia  226289468.0           65.579
17  Russian Federation  143518814.0           58.920
23              Mexico  106005199.0           72.575
25              Brazil  186127108.0           68.166

In [185]:
population_2005_3.shape

(18, 3)

In [225]:
# Choropleth Map of the World
fig = px.choropleth(population_2005_3,locations='Country Name',locationmode='country names',
                    color='Life Expectancy', color_continuous_scale='Viridis',)
fig.update_layout(title='Choropleth Map of Life Expectancy for Countries having population between 63179356 and 1147609924 in the year 2005',
                  template="simple_white")
fig.show()

## Task 1.2.4:

In [187]:
# Check for the country with the highest population in the year 2005

def find_max_country(df_merged_data):
    return df_merged_data[df_merged_data['Population'] == max(df_merged_data['Population'])].index[0]

find_max_country(df_merged_data)

print(df_merged_data.loc[find_max_country(df_merged_data)])

Country Name              China
Population         1303720000.0
Life Expectancy          71.297
Name: 134, dtype: object


In [188]:
# Create a new dataframe for the country with the highest population in the year 2005 
# (taking into consideration records from 2000 to 2010)

df_China = final_data[final_data['Country Name']=='China']
df_China.head()

Country Name Country Code  \
134        China          CHN   
301        China          CHN   
468        China          CHN   
635        China          CHN   
802        China          CHN   

                                              Geometry  Year  Life Expectancy  \
134  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2000           69.595   
301  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2001           69.981   
468  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2002           70.351   
635  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2003           70.697   
802  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2004           71.013   

       Population  
134  1.262645e+09  
301  1.271850e+09  
468  1.280400e+09  
635  1.288400e+09  
802  1.296075e+09

In [190]:
df_China.tail()

Country Name Country Code  \
1637        China          CHN   
1804        China          CHN   
1971        China          CHN   
2138        China          CHN   
2305        China          CHN   

                                               Geometry  Year  \
1637  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2006   
1804  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2007   
1971  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2008   
2138  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2009   
2305  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2010   

      Life Expectancy    Population  
1637           71.551  1.311020e+09  
1804           71.788  1.317885e+09  
1971           72.017  1.324655e+09  
2138           72.244  1.331260e+09  
2305           72.476  1.337705e+09

In [192]:
df_China.nunique(dropna=True)

Country Name        1
Country Code        1
Geometry            1
Year               11
Life Expectancy    11
Population         11
dtype: int64

In [193]:
df_China['Percentage Change'] = df_China['Life Expectancy'].pct_change()
df_China.head()

Country Name Country Code  \
134        China          CHN   
301        China          CHN   
468        China          CHN   
635        China          CHN   
802        China          CHN   

                                              Geometry  Year  Life Expectancy  \
134  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2000           69.595   
301  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2001           69.981   
468  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2002           70.351   
635  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2003           70.697   
802  MULTIPOLYGON (((109.47521 18.19770, 108.65521 ...  2004           71.013   

       Population  Percentage Change  
134  1.262645e+09                NaN  
301  1.271850e+09           0.005546  
468  1.280400e+09           0.005287  
635  1.288400e+09           0.004918  
802  1.296075e+09           0.004470

In [194]:
fig = px.bar(df_China, x='Percentage Change', y='Year',
             hover_data=['Population', 'Life Expectancy'], color='Life Expectancy',
             labels={'Percentage Change':'Percentage Change in Life Expectancy'}, height=400)
fig.show()

## Task 1.2.5: 

In [195]:
# Create a new dataframe containing the average life expectancy for each country between 2000 and 2010

df_Countries = final_data[['Country Name', 'Year', 'Geometry', 'Population', 'Life Expectancy']].groupby('Country Name',
                as_index=False).mean()
df_Countries.head()

Country Name    Population  Life Expectancy
0  Afghanistan  2.534633e+07        57.090071
1      Albania  3.004509e+06        72.697929
2      Algeria  3.326814e+07        71.694000
3       Angola  1.957829e+07        48.144429
4    Argentina  3.887895e+07        70.999929

In [227]:
# Plot a scatter plot of the average life expectancy for each country between 2000 and 2010 
# (taking the average for each country over the years)

fig = px.scatter(final_data, x='Population', y='Life Expectancy', color='Country Name', animation_frame='Year',)
fig.update_layout(title='Scatter Plot of Population VS Life Expectancy ranging from year 2000 to 2010',
                  template="plotly_dark")
fig.show()